pinot-query-runtime/src/main/java/org/apache/pinot/query/mailbox/GrpcSendingMailbox.java

```diff
@@ -319,4 +279,78 @@ static List<ByteString> toByteStrings(List<ByteBuffer> bytes, int maxByteStringS
 
     return result;
   }
+
+  private static abstract class Sender {
+    protected abstract void send(MseBlock block, List<DataBuffer> serializedStats)
+        throws IOException;
+  }
+
+  private static class SplitSender extends Sender {
+    private final GrpcSendingMailbox _mailbox;
+
+    public SplitSender(GrpcSendingMailbox mailbox) {
+      _mailbox = mailbox;
+    }
+
+    @Override
+    protected void send(MseBlock block, List<DataBuffer> serializedStats)
+        throws IOException {
+      _mailbox._statMap.merge(MailboxSendOperator.StatKey.RAW_MESSAGES, 1);
```

Comment:
nit: Given this is repeated in both senders, I don't think it belongs here. Probably we can move it to a private send method in GrpccSendingMailbox that increases stats and then call Sender.send

Reply:
I thought about that but didn't find the way to reuse the code without altering the order in which log traces and metrics get updated. But if you also saw it, I guess it is worth the change for the sake of readability. I split the code into two methods: a `send(MseBlock)` implemented by the abstract class and a `processAndSend(DataBlock)` with the specific logic for Split/NonSplit implementations.

With this approach, the `debug` trace and `SERIALIZED_BYTES` metrics are not update until the contents has been sent instead of between calculating the payload byte size and effectively send it.

Reply:
Why don't you move this whole method to `GrpcSendingMailbox`? Then each Sender would be just `processAndSend` .

Reply:
Yup, that makes sense.